File: src/main/java/com/github/kindrat/cassandra/client/util/EvenMoreFutures.java
package com.github.kindrat.cassandra.client.util;

import com.google.common.util.concurrent.FutureCallback;
import com.google.common.util.concurrent.Futures;
import com.google.common.util.concurrent.ListenableFuture;
import lombok.experimental.UtilityClass;
import lombok.extern.slf4j.Slf4j;

import java.util.concurrent.CompletableFuture;
import java.util.function.BiConsumer;
import java.util.function.Consumer;
import java.util.function.Function;

import static javafx.application.Platform.runLater;

@Slf4j
@UtilityClass
public class EvenMoreFutures {
    public static <T> CompletableFuture<T> toCompletable(ListenableFuture<T> listenableFuture) {
        CompletableFuture<T> future = new CompletableFuture<>();
        Futures.addCallback(listenableFuture, new FutureCallback<T>() {
            @Override
            public void onSuccess(T result) {
                runLater(() -> future.complete(result));
            }

            @Override
            public void onFailure(Throwable t) {
                runLater(() -> future.completeExceptionally(t));
            }
        });
        return future;
    }

    public static <O, T extends Throwable> BiConsumer<? super O, ? super T> logErrorIfPresent(String message) {
        return (obj, throwable) -> {
            if (throwable != null) {
                log.error(message, throwable);
            }
        };
    }

    public static <O, T extends Throwable> BiConsumer<? super O, ? super T> loggingConsumer(Function<O, ?> extractor) {
        return (obj, throwable) -> {
            if (throwable != null) {
                log.error(throwable.getMessage(), throwable);
            } else {
                log.info("Result : {}", extractor.apply(obj));
            }
        };
    }

    public static <O, T extends Throwable> BiConsumer<? super O, ? super T> handleErrorIfPresent(Consumer<T> handler) {
        return (obj, throwable) -> {
            if (throwable != null) {
                handler.accept(throwable);
            }
        };
    }
}

File: src/main/kotlin/com/github/kindrat/cassandra/client/util/EvenMoreFutures.kt
package com.github.kindrat.cassandra.client.util

import mu.KotlinLogging
import java.util.function.BiConsumer
import java.util.function.Consumer
import java.util.function.Function

private val logger = KotlinLogging.logger {}

object EvenMoreFutures {
    @JvmStatic
    fun <O, T : Throwable?> logErrorIfPresent(message: String?): BiConsumer<in O, in T> {
        return BiConsumer { _: O, throwable: T? -> throwable?.let { logger.error(message, it)  } }
    }

    @JvmStatic
    fun <O, T : Throwable?> loggingConsumer(extractor: Function<O, *>): BiConsumer<in O, in T> {
        return BiConsumer { obj: O, throwable: T? ->
            if (throwable != null) {
                logger.error(throwable.message, throwable)
            } else {
                logger.info("Result : ${extractor.apply(obj)}")
            }
        }
    }

    @JvmStatic
    fun <O, T : Throwable?> handleErrorIfPresent(lambda: (T) -> Unit): BiConsumer<in O, in T> {
        return BiConsumer { _: O, throwable: T? -> throwable?.let { lambda.invoke(it) } }
    }
}